dask_kubernetes/operator/controller/controller.py

```diff
@@ -525,35 +602,52 @@ async def get_desired_workers(scheduler_service_name, namespace, logger):
     # Otherwise try gracefully retiring via the RPC
     # Dask version mismatches between the operator and scheduler may cause this to fail in any number of unexpected ways
     try:
-        comm_address = await get_scheduler_address(
+        comm_address = await get_scheduler_address(  # type: ignore[no-untyped-call]
             scheduler_service_name,
             namespace,
             allow_external=False,
         )
-        async with rpc(comm_address) as scheduler_comm:
+        async with rpc(comm_address) as scheduler_comm:  # type: ignore[no-untyped-call]
             return await scheduler_comm.adaptive_target()
     except Exception as e:
         raise SchedulerCommError(
             "Unable to get number of desired workers from scheduler"
         ) from e
 
 
-worker_group_scale_locks = defaultdict(lambda: asyncio.Lock())
+worker_group_scale_locks: dict[str, asyncio.Lock] = defaultdict(lambda: asyncio.Lock())
 
 
 @kopf.on.field("daskcluster.kubernetes.dask.org", field="spec.worker.replicas")
 async def daskcluster_default_worker_group_replica_update(
-    name, namespace, old, new, **kwargs
-):
+    name: str | None,
+    namespace: str | None,
+    old: Any | None,
+    new: Any | None,
+    **__: Any,
+) -> None:
+    if name is None or namespace is None:
+        raise kopf.PermanentError("Name or Namespace not given.")
```

Comment:
Are there really cases where `kopf` would call this with `None` values? I'm not sure we should be so aggressive with checking hooks from dependencies like this.

Reply:
I'm actually not too familiar with `kopf` and what a circumstance is where `name` or `namespace` could be `None` (probably some error cases where *something* still wants to be cleaned up or at least logged).

The test for the correct Callable is defined here in kopf: https://github.com/nolar/kopf/blob/9243c8fbb2a36c1b7cad1f65a647d3b9876a5b17/kopf/_core/intents/callbacks.py#L109

As you can see, it expects any "on change" Callable to be typed that it also could expect `None` for `name` or `namespace`. Because Callables are contravariant, giving it a Callable where `name` can't be `None` will result in a `mypy` error on applying the decorator `kopf.on.[...]` to this function here.
The other solution would be to leave out checking for `None` *inside* the function, but then every access to these variables would result in `mypy` errors since they potentially could be `None`.

Maybe we could replace this with an `assert`, then it at least would be a bit less visually verbose?

Reply:
Adding assertions where we make assumptions like this make the most sense to me. Thanks!

Reply:
Added `asserts` now.